PXC-3352 [5.7]: Unexpected ERROR 1205 modifying a child table in a FK relationship

venkatesh-prasad-v opened the pull request:
PXC-3352: Unexpected ERROR 1205 modifying a child table in a FK relationship

https://jira.percona.com/browse/PXC-3352

Problem
-------
When deleting/updating from a child table in a FK relationship, if the
parent table's referenced rows are locked, the operation on a child table
failed with lock wait timeout error when the parent table is unlocked.

Analysis
--------
1. For DELETE query

This is the stacktrace of the DELETE query while waiting for the parent
rows to be unlocked.
```
__pthread_cond_wait
os_event::wait
os_event::wait_low
os_event_wait_low
lock_wait_suspend_thread
row_ins_check_foreign_constraint
wsrep_row_upd_check_foreign_constraint 
row_upd_sec_index_entry
row_upd_sec_step
row_upd
row_upd_step
row_update_for_mysql_using_upd_graph
row_update_for_mysql
```
However, the same DELETE query doesn't wait in PS-5.7. 

In PS-5.7, when a row is being modified by an UPDATE/DELETE query, it even
updates the secondary index if there any index defined on the table. While
doing that, it first marks the row as deleted in the secondary index and
inserts into the index with the new value if it is an UPDATE query. This
updation of secondary index is necessary only for UPDATE queries and there
is no need to do the same for DELETE queries.

However, the same was not being done properly in PXC-5.7 because of the
special handling in WSREP for the foreign keys in InnoDB. This made even
the DELETE query to inadvertently update the secondary index
(row_ins_check_foreign_constraint()) thereby trying to lock the parent
table and thus causing lock wait timeout error.

2. For UPDATE query

During Update/Delete, whenever we fail to acquire lock on a row, the
transaction status is to DB_LOCK_WAIT and transaction enters into wait
state waiting for the row to be unlocked; and when the row is unlocked, the
operation is retried.

However, in case of PXC, the temporary DB_LOCK_WAIT during foreign key
checks was being translated into a permanent DB_LOCK_WAIT_TIMEOUT because
of the fallthrough in the error handling code thereby causing the queyr to
fail without retrying the operation.


Fix
---
1. The server is made to behave same way as of the PS-5.7 by removing the
check for DELETE query in `wsrep_row_upd_check_foreign_constraints()`
function which caused the thread to enter the wait.

2. DB_LOCK_WAIT error is now properly handled and the server retries the
operation.

Additionally, this patch 

1. Adds new foreign key tests in galera suite invoking their respective
test from InnoDB with galera enabled.
2. Re-enables the disabled `galera_fk_multitable` and
`galera_toi_ddl_fk_insert` test cases.

Testing
---
v1 - Jenkins: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/51/testReport/
v2 - Jenkins: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/57/testReport/


Note to Reviewers:
---
* Many tests in the above link are failing because the galera suite was run with higher value of `--parallel` causing making ports to be unavailable during the test run. To address this issue, [PXC-3364](https://jira.percona.com/browse/PXC-3364) has been reported.

altmannmarcelo commented on storage/innobase/row/row0upd.cc:
```diff
@@ -420,10 +420,8 @@ wsrep_row_upd_check_foreign_constraints(
-			|| row_upd_changes_first_fields_binary(
-				entry, index, node->update,
-				foreign->n_fields))) {
+		    && row_upd_changes_first_fields_binary(
```
I'm trying to understand why node->is_delete is not required here. I can see that on `row_upd_sec_index_entry` we call `row_upd_index_is_referenced` which is the base function for `wsrep_row_upd_check_foreign_constraints`, then if `WITH_WSREP` is defined we call `wsrep_row_upd_check_foreign_constraints` which basically runs a check from `row_ins_check_foreign_constraint` on the parent table instead of on the child table (calling `row_ins_check_foreign_constraint` with 1st parameter = TRUE). I have two questions:

1. Why this second call to check FK constrain is required only on WSREP
2. By removing `node->is_delete` we won't be calling the function that is causing the issue, what happens on an update that returns true for  row_upd_changes_first_fields_binary and has the parent table records locked?

altmannmarcelo replied:
While talking with @venkatesh-prasad-v about this fix we found out that test case is failing for UPDATES:

```
setup
CREATE TABLE parent_t ( id INT AUTO_INCREMENT PRIMARY KEY );
CREATE TABLE child_t ( id INT AUTO_INCREMENT PRIMARY KEY, parent_id INT,
CONSTRAINT `FK_child_parent_id` FOREIGN KEY (`parent_id`) REFERENCES `parent_t` (`id`));
INSERT INTO parent_t VALUES (), (), (), (), ();
INSERT INTO child_t (parent_id) SELECT id FROM parent_t;
Session 1
BEGIN;
SELECT * FROM parent_t WHERE id = 1 FOR UPDATE;
Session 2
UPDATE child_t SET parent_id = 2 WHERE parent_id = 1;
Session 1
COMMIT;--(ROLLBACK;, abort, doesn't matter)
Session 2
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
```

venkatesh-prasad-v replied:
Hi Marcelo,

Thank you for your time on reviewing this.

> 1. Why this second call to check FK constrain is required only on WSREP

The call to `row_upd_index_is_referenced()` function is for checking if there are any references to the index in the current table. i.e in simple words, it only means that if there are any child tables referring to the current table object.

Whereas the the call to `wsrep_row_upd_index_is_foreign()` function is for checking if the current table has a any foreign key defined.

Both these calls required since they will is used by the `wsrep_append_foreign_key` for generating writesets for the update operation.

> 2. By removing `node->is_delete` we won't be calling the function that is causing the issue, what happens on an update that returns true for row_upd_changes_first_fields_binary and has the parent table records locked?

In such cases, updates also wait for the lock on the parent row to be released. However there was an issue in the error handling code that made the UPDATE queries to fail even after the parent rows were released. This has been fixed in the latest version of the patch.

altmannmarcelo reviewed (commented):
@venkatesh-prasad-v I have a few questions around the part of code you changed.

Also, not mandatory since its still a WIP but have a read at https://confluence.percona.com/pages/viewpage.action?pageId=68393728 (especially the part that mentions inserting the link for JIRA issue for easy access).

altmannmarcelo reviewed (changes requested):
While talking with @venkatesh-prasad-v about this fix we found out that test case is failing for UPDATES:

```
setup
CREATE TABLE parent_t ( id INT AUTO_INCREMENT PRIMARY KEY );
CREATE TABLE child_t ( id INT AUTO_INCREMENT PRIMARY KEY, parent_id INT,
CONSTRAINT `FK_child_parent_id` FOREIGN KEY (`parent_id`) REFERENCES `parent_t` (`id`));
INSERT INTO parent_t VALUES (), (), (), (), ();
INSERT INTO child_t (parent_id) SELECT id FROM parent_t;
Session 1
BEGIN;
SELECT * FROM parent_t WHERE id = 1 FOR UPDATE;
Session 2
UPDATE child_t SET parent_id = 2 WHERE parent_id = 1;
Session 1
COMMIT;--(ROLLBACK;, abort, doesn't matter)
Session 2
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
```

venkatesh-prasad-v commented:
> While talking with @venkatesh-prasad-v about this fix we found out that test case is failing for UPDATES:
> 
> ```
> setup
> CREATE TABLE parent_t ( id INT AUTO_INCREMENT PRIMARY KEY );
> CREATE TABLE child_t ( id INT AUTO_INCREMENT PRIMARY KEY, parent_id INT,
> CONSTRAINT `FK_child_parent_id` FOREIGN KEY (`parent_id`) REFERENCES `parent_t` (`id`));
> INSERT INTO parent_t VALUES (), (), (), (), ();
> INSERT INTO child_t (parent_id) SELECT id FROM parent_t;
> Session 1
> BEGIN;
> SELECT * FROM parent_t WHERE id = 1 FOR UPDATE;
> Session 2
> UPDATE child_t SET parent_id = 2 WHERE parent_id = 1;
> Session 1
> COMMIT;--(ROLLBACK;, abort, doesn't matter)
> Session 2
> ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
> ```

This was happening because, the the temporary DB_LOCK_WAIT error was being translated to DB_LOCK_WAIT_TIMEOUT there by causing the UPDATE command to fail with the timeout error. This has been fixed in the latest revision of the patch.

```
2504                     }
2505                     break;
2506                 case DB_LOCK_WAIT_TIMEOUT:
2507                 case DB_LOCK_WAIT:
2508                     err = DB_LOCK_WAIT_TIMEOUT;
2509                     break;
2510                 default:
```

altmannmarcelo commented:
The new path looks good to me. UPDATE test is passing now.

venkatesh-prasad-v commented:
NOTE
---
While discussing with @altmannmarcelo, we found a case where some update queries are immediate on PS and the same wait in PXC.

The issue is that if the parent table has referenced row(s) locked, then the update query on a child table will wait until the parent row is unlocked and this diverges from the PS behavior where the such updates succeed immediately.
```
CREATE TABLE parent_t (id INT AUTO_INCREMENT PRIMARY KEY )ENGINE = INNODB;
CREATE TABLE child_t (id INT AUTO_INCREMENT PRIMARY KEY, parent_id INT,
                      CONSTRAINT `FK_child_parent_id` FOREIGN KEY (`parent_id`)
                      REFERENCES `parent_t` (`id`)) ENGINE = INNODB;

INSERT INTO parent_t VALUES (), (), (), (), ();
INSERT INTO child_t (parent_id) SELECT id FROM parent_t;

--con1
BEGIN;
SELECT * FROM parent_t WHERE id = 1 FOR UPDATE;

--con2
UPDATE child_t SET parent_id = 1 WHERE id = 1; - Both Succeed
UPDATE child_t SET parent_id = 2 WHERE id = 1; - PXC Waits
UPDATE child_t SET parent_id = 3 WHERE id = 1; - PXC Waits
UPDATE child_t SET parent_id = 4 WHERE id = 1; - PXC Waits
UPDATE child_t SET parent_id = 5 WHERE id = 1; - PXC Waits
```

However, this is not in the context of the current bug and shall be handled as part of https://jira.percona.com/browse/PXC-3374.

Commit:
PXC-3352: Unexpected ERROR 1205 modifying a child table in a FK relationship

https://jira.percona.com/browse/PXC-3352

Problem
-------
When deleting/updating from a child table in a FK relationship, if the
parent table's referenced rows are locked, the operation on a child table
failed with lock wait timeout error when the parent table is unlocked.

Analysis
--------
1. For DELETE query

This is the stacktrace of the DELETE query while waiting for the parent
rows to be unlocked.

__pthread_cond_wait
os_event::wait
os_event::wait_low
os_event_wait_low
lock_wait_suspend_thread
row_ins_check_foreign_constraint
wsrep_row_upd_check_foreign_constraint
row_upd_sec_index_entry
row_upd_sec_step
row_upd
row_upd_step
row_update_for_mysql_using_upd_graph
row_update_for_mysql

However, the same DELETE query doesn't wait in PS-5.7.

In PS-5.7, when a row is being modified by an UPDATE/DELETE query, it even
updates the secondary index if there any index defined on the table. While
doing that, it first marks the row as deleted in the secondary index and
inserts into the index with the new value if it is an UPDATE query. This
updation of secondary index is necessary only for UPDATE queries and there
is no need to do the same for DELETE queries.

However, the same was not being done properly in PXC-5.7 because of the
special handling in WSREP for the foreign keys in InnoDB. This made even
the DELETE query to inadvertently update the secondary index
(row_ins_check_foreign_constraint()) thereby trying to lock the parent
table and thus causing lock wait timeout error.

2. For UPDATE query

During Update/Delete, whenever we fail to acquire lock on a row, the
transaction status is to DB_LOCK_WAIT and transaction enters into wait
state waiting for the row to be unlocked; and when the row is unlocked, the
operation is retried.

However, in case of PXC, the temporary DB_LOCK_WAIT during foreign key
checks was being translated into a permanent DB_LOCK_WAIT_TIMEOUT because
of the fallthrough in the error handling code thereby causing the queyr to
fail without retrying the operation.

Fix
---
1. The server is made to behave same way as of the PS-5.7 by removing the
check for DELETE query in `wsrep_row_upd_check_foreign_constraints()`
function which caused the thread to enter the wait.

2. DB_LOCK_WAIT error is now properly handled and the server retries the
operation.

Additionally, this patch

1. Adds new foreign key tests in galera suite invoking their respective
test from InnoDB with galera enabled.
2. Re-enables the disabled `galera_fk_multitable` and
`galera_toi_ddl_fk_insert` test cases.
